File: src/main/cpp/lib/TaxRule.cpp
/*******************************************************************************
 *
 * Copyright (c) {2003-2019} Murex S.A.S. and its affiliates.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 *******************************************************************************/

#include "finance/TaxRule.h"
#include <map>
#include <string>
#include "domain/book/Novel.h"
#include <typeinfo>

namespace
{
	std::map<std::string, double> TAX_RATES =
	{
		{"USA", 1.15},
		{"France", 1.25},
		{"UK", 1.20},
		{"Spain", 1.10},
		{"China", 1.35},
		{"Japan", 1.30},
		{"Australia", 1.13},
		{"Germany", 1.22},
	};


	double getTaxRate(const std::string& country)
	{
		return TAX_RATES[country];
	}
}

double finance::getApplicableRate(const domain::country::Country& invoiceCountry, const domain::book::Book& book) {
	if (invoiceCountry.getName() == "Germany") {
		if (book.getAuthor().getNationality().getName() == "Germany") {
			return 1.05;
		}
	}
	//FIXME does this typeid(thing) work?
	if (invoiceCountry.getName() == "USA") {
		if (typeid(book).name() == std::string("class domain::book::Novel")) {
			return getTaxRate(invoiceCountry.getName()) * 0.98;
		}
	}
	if (invoiceCountry.getName() == "UK") {
		if (typeid(book).name() == std::string("class domain::book::Novel")) {
			return getTaxRate(invoiceCountry.getName()) * 0.93;
		}
	}
	if (invoiceCountry.getName() == "China") {
		if (book.getLanguage() != domain::country::MANDARIN) {
			return 1;
		}
	}
	if (invoiceCountry.getName() == "Spain") {
		if (book.getLanguage() != domain::country::SPANISH) {
			return 1;
		}
	}
	return getTaxRate(invoiceCountry.getName());
}

File: csharp/Application/Finance/TaxRule.cs
/*******************************************************************************
 *
 * Copyright (c) {2003-2019} Murex S.A.S. and its affiliates.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the MIT License
 * which accompanies this distribution, and is available at
 * https://opensource.org/licenses/MIT
 *
 *******************************************************************************/

using System.Collections.Generic;
using System.Collections.ObjectModel;
using Application.Domain.Book;
using Application.Domain.Country;
using static Application.Domain.Country.Language;

namespace Application.Finance
{
    public static class TaxRule
    {
        private static readonly IReadOnlyDictionary<string, double> TaxRates =
            new ReadOnlyDictionary<string, double>(
                new Dictionary<string, double>
                {
                    {"USA", 1.15},
                    {"France", 1.25},
                    {"UK", 1.20d},
                    {"Spain", 1.10d},
                    {"China", 1.35d},
                    {"Japan", 1.30d},
                    {"Australia", 1.13d},
                    {"Germany", 1.22d},
                });

        public static double GetApplicableRate(Country invoiceCountry, IBook book)
        {
            if (invoiceCountry.Name.Equals("Germany"))
                if (book.Author.Nationality.Name.Equals("Germany"))
                    return 1.05;

            if (invoiceCountry.Name.Equals("USA"))
                if (book is Novel)
                    return GetTaxRate(invoiceCountry.Name) * 0.98;

            if (invoiceCountry.Name.Equals("UK"))
                if (book is Novel)
                    return GetTaxRate(invoiceCountry.Name) * 0.93;


            if (invoiceCountry.Name.Equals("China"))
                if (!book.Language.Equals(Mandarin))
                    return 1;

            if (invoiceCountry.Name.Equals("Spain"))
                if (!book.Language.Equals(Spanish))
                    return 1;

            return GetTaxRate(invoiceCountry.Name);
        }

        private static double GetTaxRate(string country) => TaxRates[country];
    }
}
